src/main/networking/mac-ca.ts

```diff
@@ -0,0 +1,175 @@
+/**
+ * The depInjection interface is used for testing.
+ */
+interface depInjection {
+  getPEMCertificates?: (workdir: string, keychain?: string) => AsyncIterable<string>;
+  spawnFile?: (_1: string, _2: string[], _3?: unknown) => Promise<{stdout: string}>;
+  listKeychains?: () => AsyncIterable<string>;
+  getFilteredCertificates?: (workdir: string, keychain: string) => AsyncIterable<string>;
+}
+
+/**
+ * Asynchronously enumerate the certificate authorities that should be used to
+ * build the Rancher Desktop trust store, in PEM format in undefined order.
+ */
+export default async function* getMacCertificates(deps_: depInjection = {}): AsyncIterable<string> {
+  const deps = _.defaults(deps_, { listKeychains, getFilteredCertificates }) as Required<depInjection> ;
+  const workdir = await fs.promises.mkdtemp('rancher-desktop-certificates-');
+
+  try {
+    for await (const keychain of deps.listKeychains()) {
+      yield * deps.getFilteredCertificates(workdir, keychain);
+    }
+  } finally {
+    await fs.promises.rm(workdir, {
+      recursive: true, force: true, maxRetries: 3
+    });
+  }
+}
+
+/**
+ * Return all keychains that we should import from.
+ */
+export async function* listKeychains(): AsyncIterable<string> {
+  const { stdout } = await spawnFile('/usr/bin/security', ['list-keychains'],
+    { stdio: ['ignore', 'pipe', console] });
+
+  for (const line of stdout.split(/\n/)) {
+    yield line.trim().replace(/^"|"$/g, '');
+  }
+  try {
+    // Add the system root certificates keychain; this is normally not listed
+    // as it wouldn't include _client_ certificates.
+    const rootCerts = '/System/Library/Keychains/SystemRootCertificates.keychain';
+
+    await fs.promises.access(rootCerts, fs.constants.R_OK);
+    yield rootCerts;
+  } catch (ex) { /* swallow the error */ }
+}
+
+/**
+  * Asynchronously enumerate PEM-encoded certificates from the given keychain in
+  * undefined order.
+  *
+  * @param workdir A temporary directory where files can be written.
+  * @param keychain The full path to the keychain database to enumerate.
+  * @note This is only exported for testing.
+  */
+export async function* getFilteredCertificates(workdir: string, keychain: string, deps_: depInjection = {}): AsyncIterable<string> {
+  console.debug(`getting certificates from ${ keychain }...`);
+
+  const deps = _.defaults(deps_, { getPEMCertificates, spawnFile }) as Required<depInjection>;
+  const certIterator = deps.getPEMCertificates(workdir, keychain);
+
+  for await (const certPEM of certIterator) {
+    const cert = new crypto.X509Certificate(certPEM);
+    const certPath = path.join(workdir, 'cert.pem');
+
+    console.error('got cert', cert, cert.ca, cert.issuer, cert.subject, certPEM);
+    if (!cert.ca) {
+      console.debug('Skipping non-CA certificate', cert.subject);
+      continue;
+    }
+    if (cert.issuer !== cert.subject) {
+      console.debug('Skipping non-self-signed certificate', cert.subject);
+      continue;
+    }
```

Comment:
Why are you skipping those?

Reply:
It might be a good idea to document why this function skips certain certs, or which certs we're interested in to provide some insight into why we would skip.

Reply:
Hmm, you're right that we shouldn't skip this (it would have skipped Let's Encrypt's CA cert, I think); dropped the requirement.